tools/keycloak/sync/SyncFactory.cs

```diff
@@ -384,24 +383,34 @@ private async Task SyncUsersAsync(StringBuilder log, StringBuilder errorLog)
+                $"{_keycloakManagementClient.GetIntegrationEnvUri()}/users/{Uri.EscapeDataString(username)}/roles",
+                (response) => response.StatusCode == HttpStatusCode.NotFound)).Data;
 
-            IEnumerable<UserRoleOperation> userRolesToAdd = user.UserRoles.Where(ur => kUserRoles.Roles.All(kr => kr.Name != ur.Role.Name)).Select(ur => new UserRoleOperation() { RoleName = ur.Role.Name, Username = username, Operation = "add" });
-            IEnumerable<UserRoleOperation> userRolesToRemove = kUserRoles.Roles.Where(kur => user.UserRoles.All(ur => ur.Role.Name != kur.Name)).Select(kur => new UserRoleOperation() { RoleName = kur.Name, Username = username, Operation = "del" });
-            var allUserRoleMappings = userRolesToAdd.Concat(userRolesToRemove);
+            IEnumerable<RoleModel> userRolesToAdd = user.UserRoles.Where(ur => kUserRoles.All(kr => kr.Name != ur.Role.Name && ur.Role.IsDisabled == false)).Select(ur => new RoleModel() { Name = ur.Role.Name });
+            IEnumerable<RoleModel> userRolesToRemove = kUserRoles.Where(kur => user.UserRoles.All(ur => ur.Role.Name != kur.Name)).Select(ur => new RoleModel() { Name = ur.Name });
 
-            foreach (UserRoleOperation operation in allUserRoleMappings)
+            if (userRolesToAdd.Any())
             {
-                _logger.LogInformation($"Executing operation '{operation.Operation}' on role '{operation.RoleName}' to user '{operation.Username}'");
+                _logger.LogInformation($"Executing operation 'add' on roles '{string.Join(',', userRolesToAdd.Select(r => r.Name))}' to user '{username}'");
+                var response = await _keycloakManagementClient.SendJsonAsync($"{_keycloakManagementClient.GetIntegrationEnvUri()}/users/{Uri.EscapeDataString(username)}/roles", HttpMethod.Post, userRolesToAdd);
+                if (!response.IsSuccessStatusCode)
+                {
+                    throw new HttpClientRequestException(response, $"Failed to update the user role mappings for '{username}' during operation 'add' on roles '{string.Join(',', userRolesToAdd.Select(r => r.Name))}'");
+                }
+                LogInfo(log, $"Executed operation 'add' on roles '{string.Join(',', userRolesToAdd.Select(r => r.Name))}' to user '{username}'");
+            }
+
+            foreach (RoleModel userRoleToRemove in userRolesToRemove)
+            {
+                _logger.LogInformation($"Executing operation 'delete' on role '{userRoleToRemove.Name}' to user '{username}'");
 
-                var response = await _keycloakManagementClient.SendJsonAsync($"{_keycloakManagementClient.GetIntegrationEnvUri()}/user-role-mappings", HttpMethod.Post, operation);
+                var response = await _keycloakManagementClient.SendAsync($"{_keycloakManagementClient.GetIntegrationEnvUri()}/users/{Uri.EscapeDataString(username)}/roles/{Uri.EscapeDataString(userRoleToRemove.Name)}", HttpMethod.Delete);
```

Comment:
nit: there seems to be a specialized "DeleteAsync" method - wouldn't it make sense to use that over the generic SendAsync?